src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+#include <headerssync.h>
+#include <logging.h>
+#include <util/check.h>
+#include <pow.h>
+#include <timedata.h>
+
+// Store a commitment to a header every HEADER_COMMITMENT_FREQUENCY blocks
+constexpr size_t HEADER_COMMITMENT_FREQUENCY = 571;
+// Pull out headers for acceptance during redownload once our buffer exceeds this amount
+constexpr size_t REDOWNLOAD_HEADERS_THRESHOLD = 25*HEADER_COMMITMENT_FREQUENCY;
+// Ensure that the redownload buffer is a multiple of the header commitment
+// frequency, so that we aren't wasting memory.
+static_assert(REDOWNLOAD_HEADERS_THRESHOLD % HEADER_COMMITMENT_FREQUENCY == 0);
+
+// Our memory analysis assumes 48 bytes for a CompressedHeader (so we should
+// re-calculate parameters if we compress further)
+static_assert(sizeof(CompressedHeader) == 48);
+
+HeadersSyncState::HeadersSyncState(NodeId id, const Consensus::Params& consensus_params) :
+    m_id(id), m_consensus_params(consensus_params)
+{ }
+
+// Free any memory in use, and mark this object as no longer usable.
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

Perhaps document that this is required to guarantee that the object isn't reused with the same SaltedTxidHasher for another sync attempt.

Reply:
Done.